File: CarePackage/Software/Pcsx2.cs
namespace CarePackage.Software;

public class Pcsx2 : BaseSoftware
{
    public override string           Key            { get; set; } = "pcsx2-stable";
    public override string           Name           { get; set; } = "PCSX2";
    public override SoftwareCategory Category       { get; set; } = SoftwareCategory.Gaming;
    public override string           DownloadName   { get; set; } = "PCSX2.7z";
    public override bool             IsArchive      { get; set; } = true;
    public override bool             ShouldCacheUrl { get; set; } = true;
    public override bool             RequiresAdmin  { get; set; } = false;
    public override Bitmap           Icon           { get; set; } = Resources.Icons.pcsx2;
    public override string           Homepage       { get; set; } = "https://pcsx2.net";
    
    private readonly GitHubService _github;

    public Pcsx2(IServiceProvider services)
    {
        _github = services.GetRequiredService<GitHubService>();
    }
    
    public override async Task<string> GetDownloadUrlAsync(CancellationToken ct)
    {
        var assets = await _github.GetLatestRepositoryReleaseAssetsAsync("pcsx2", "pcsx2", ct);
        var asset  = assets.FirstOrDefault(a => a.EndsWith("windows-x64-Qt.7z"));
        
        DownloadUrlResolveException.ThrowIf(asset is null);

        return asset;
    }
}

File: packages/app/src/services/software/definitions/pcsx2.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory } from 'shared';
import { GithubReleasesScraper } from '../githubReleasesScraper';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'pcsx2-stable';
	public name = 'PCSX2';
	public category = [SoftwareCategory.Emulation, SoftwareCategory.Gaming];
	public downloadName = 'PCSX2.7z';
	public isArchive = true;
	public shouldCacheUrl = true;
	public requiresAdmin = false;
	public deprecated = false;
	public icon = 'pcsx2.png';
	public homepage = 'https://pcsx2.net';

	public async resolveDownloadUrl() {
		const owner   = 'pcsx2';
		const repo    = 'pcsx2';
		const release = await GithubReleasesScraper.getRelease(owner, repo, r => r.endsWith('windows-x64-Qt.7z'));
		if (release.isErr()) {
			return err(release.error);
		}

		return ok(release.value);
	}
}
